cmd/dirrotate.cpp

```diff
@@ -0,0 +1,162 @@
+
+  DESCRIPTION
+  + "The primary use case of this command is to find,"
+    " for a given basis direction set,"
+    " an appropriate rotation that preserves the homogeneity of coverage on the sphere"
+    " but that minimises the maximal peak amplitude along the physical axes of the scanner,"
+    " so as to minimise the peak gradient system demands."
+    " It can alternatively be used to introduce a random rotation"
+    " to hopefully prevent any collinearity between directions in different shells,"
+    " by requesting only a single permutation.";
+
+  ARGUMENTS
+    + Argument ("in", "the input direction file").type_file_in()
+    + Argument ("out", "the output direction file").type_file_out();
+
+
+  OPTIONS
+    + Option ("permutations", "number of permutations to try"
+                              " (default: " + str(default_permutations) + ")")
+    +   Argument ("num").type_integer(1)
+
+    + Option ("cartesian", "Output the directions in Cartesian coordinates [x y z]"
+                           " instead of [az el].");
+}
+// clang-format on
+
+using axis_type = Eigen::Matrix<default_type, 3, 1>;
+using rotation_type = Eigen::AngleAxis<default_type>;
+using rotation_transform_type = Eigen::Transform<default_type, 3, Eigen::Isometry>;
+
+class Shared {
+public:
+  Shared(const Eigen::MatrixXd &directions, size_t target_num_permutations)
+      : directions(directions),
+        target_num_permutations(target_num_permutations),
+        num_permutations(0),
+        progress(target_num_permutations == 1 ? "randomising direction set orientation" : "optimising directions for peak gradient load", target_num_permutations),
+        best_rotation(0.0, axis_type{0.0, 0.0, 0.0}),
+        min_peak(1.0),
+        original_peak(directions.array().abs().maxCoeff()) {}
+
+  bool update(default_type peak, const rotation_type &rotation) {
+    std::lock_guard<std::mutex> lock(mutex);
+    if (peak < min_peak) {
+      min_peak = peak;
+      best_rotation = rotation;
+      progress.set_text(
+          std::string(target_num_permutations == 1 ? "randomising direction set orientation" : "optimising directions for peak gradient load")
+          + " (original = " + str(original_peak, 6) + "; "
+          + (target_num_permutations == 1 ? "rotated" : "best")
+          + " = " + str(min_peak, 6) + ")");
+    }
+    ++num_permutations;
+    ++progress;
+    return num_permutations < target_num_permutations;
+  }
+
+  default_type peak(const rotation_type &rotation) const {
+    return (rotation_transform_type(rotation).linear() * directions.transpose()).array().abs().maxCoeff();
+  }
+
+  const rotation_type &get_best_rotation() const { return best_rotation; }
+
+protected:
+  const Eigen::MatrixXd &directions;
+  const size_t target_num_permutations;
+  size_t num_permutations;
+  ProgressBar progress;
+  rotation_type best_rotation;
+  default_type min_peak;
+  default_type original_peak;
+  std::mutex mutex;
```

Comment:
We can probably use `MutexProtected` (in `core/mutexprotected.h`) here?

Reply:
I created this command by duplicating `dirflip`. On `dev` that's currently not using that class. So should modify both.

Reply:
In this scenario, there's *two* member variables that should be checked and updated within a lock: `best_rotation` and `min_peak`. I *think* that given it's a full lock rather than an atomic with limited memory ordering, just locking one and then modifying the other while the lock is held shouldn't cause any memory incongruency issues. However given the design of `MutexProtected` I think the better solution is to define a `struct` that encompasses both the rotation and the resulting minimal peak value, and then that becomes the templated object within the `MutexProtected` class.

(This logic potentially extends to other places where `MutexProtected` could be used; essentially a form of "shared" class that encompasses everything controlled under a singular lock)

Reply:
Yes, I agree that using a `struct ProtectedFields{...}` is the most natural way of using `MutexProtected` in this case.

Reply:
Resolved by 07693ac799ae75e8ba97bf95631e82aafa16c2cd.